File: c/fullpath.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <stdbool.h>

typedef struct invocation {
  int num_args;
  char **args;

  char *cwd;

  int num_chars;
  int num_paths;
  bool free_paths;
  char **relative_paths;

  bool copy_result;
  bool print_help;

  pid_t child_pid;
} Invocation;

void print_help() {
  printf("usage: fullpath *[relative-paths] [-c]\n");
  printf("\n");
  printf("  Prints the fullpath of the paths\n");
  printf("  If no paths are given as args, it will read them from stdin\n");
  printf("\n");
  printf("  If there is only one path, the trailing newline is omitted\n");
  printf("\n");
  printf("  The -c flag will copy the results into your pasteboard\n");
}

void output(FILE *stream, Invocation *invocation) {
  if(invocation->num_paths == 1)
    fprintf(stream, "%s/%s", invocation->cwd, invocation->relative_paths[0]);
  else
    for(int i=0; i < invocation->num_paths; ++i)
      fprintf(stream, "%s/%s\n", invocation->cwd, invocation->relative_paths[i]);
}


void remove_empties(char **strings, int *num_strings) {
  int from, to;
  for(from=0, to=0; from<*num_strings; ++from)
    if(*strings[from])
      strings[to++] = strings[from];
  *num_strings = to;
}

void remove_string(char **strings, int*num_strings, char *to_remove, bool *removed) {
  int from=0, to=0;
  for(; from<*num_strings; ++from)
    if(0 != strcmp(strings[from], to_remove))
      strings[to++] = strings[from];
    else
      *removed = true;
  *num_strings = to;
}

void set_invocation_counts(struct invocation *invcn, char **strings, int num_strings) {
  for(int istring=0; istring < num_strings; ++istring, ++invcn->num_chars, ++invcn->num_paths)
    for(int ichar=0; strings[istring][ichar]; ++ichar) {
      ++invcn->num_chars;
      if(strings[istring][ichar] == '\n')
        ++invcn->num_paths;
    }
}

void read_paths(FILE *stream, Invocation *invocation) {
  // read lines into a list since we don't know how many there are going to be
  struct string_list {
    char* string;
    struct string_list* next;
  };

  struct string_list *head=NULL, *tail=NULL;
  invocation->num_paths = 0;
  size_t size = 0;
  char* line;
  while(1) {
    // setting line to NULL causes getline to allocate a new buffer
    // note that it's our responsibility to free this memory
    line = NULL;
    int result = getline(&line, &size, stream);
    if(result == -1)
      break;
    else if (0 < result && line[result-1] == '\n')
      line[result-1] = '\0';
    struct string_list* node = (struct string_list*)malloc(sizeof(struct string_list));
    node->string = line;
    node->next   = NULL;
    ++(invocation->num_paths);
    if(!head) head       = node;
    if(tail)  tail->next = node;
    tail = node;
  }

  invocation->relative_paths = (char**)malloc(invocation->num_paths * sizeof(char*));
  invocation->free_paths     = true;

  for(int i=0; i < invocation->num_paths; i++) {
    invocation->relative_paths[i] = head->string;
    struct string_list* next_head = head->next;
    free(head);
    head = next_head;
  }
}

int main(int argc, char **argv) {
  // invocation info
  struct invocation invcn;
  invcn.num_args       = argc-1;
  invcn.args           = argv+1;

  char cwd[1024];
  invcn.cwd            = cwd;
  getcwd(invcn.cwd, sizeof(cwd));

  invcn.num_chars      = 0;
  invcn.num_paths      = 0;
  invcn.free_paths     = false;
  invcn.relative_paths = 0; // filled in later

  invcn.copy_result    = false;
  invcn.print_help     = false;
  invcn.child_pid      = 0;
  set_invocation_counts(&invcn, argv+1, argc-1);


  // break newlines (omg, this is misery >.<)
  char *all_args       = (char*)malloc(invcn.num_chars*sizeof(char));
  invcn.relative_paths = (char**)malloc(invcn.num_paths*sizeof(char*));

  invcn.relative_paths[0] = all_args;
  for(int i_from_str=0, i_to_str=1, i_all_args=0; i_from_str < invcn.num_args; ++i_from_str) {
    char *from_str = invcn.args[i_from_str];
    for(int ichar=0; from_str[ichar]; ++ichar)
      if(from_str[ichar] == '\n') {
        all_args[i_all_args++] = '\0';
        invcn.relative_paths[i_to_str++] = all_args+i_all_args;
      } else {
        all_args[i_all_args++] = from_str[ichar];
      }
    all_args[i_all_args++] = '\0';
    invcn.relative_paths[i_to_str++] = all_args+i_all_args;
  }

  // remove empty strings and flags
  remove_empties(invcn.relative_paths, &invcn.num_paths);
  remove_string(invcn.relative_paths, &invcn.num_paths, "-h",     &invcn.print_help);
  remove_string(invcn.relative_paths, &invcn.num_paths, "--help", &invcn.print_help);
  remove_string(invcn.relative_paths, &invcn.num_paths, "-c",     &invcn.copy_result);
  remove_string(invcn.relative_paths, &invcn.num_paths, "--copy", &invcn.copy_result);

  if(invcn.print_help) {
    print_help();
    goto done;
  }

  if(!invcn.num_paths) {
    free(invcn.relative_paths);
    read_paths(stdin, &invcn);
    remove_empties(invcn.relative_paths, &invcn.num_paths);
  }

  if(invcn.copy_result) {
    int fdescs[2]; // file descriptors
    pipe(fdescs);
    if(!(invcn.child_pid=fork())) {
      dup2(fdescs[0], STDIN_FILENO);
      close(fdescs[0]);
      close(fdescs[1]);
      char* pbcopy_argv[] = {"pbcopy", 0};
      execvp(pbcopy_argv[0], pbcopy_argv);
    } else {
      FILE* write_stream = fdopen(fdescs[1], "w");
      output(write_stream, &invcn);
      fclose(write_stream);
      close(fdescs[0]);
      close(fdescs[1]);
    }
  }
  output(stdout, &invcn);

done:
  if(invcn.free_paths) {
    for(int i=0; i < invcn.num_paths; ++i)
      free(invcn.relative_paths[i]);
  }
  free(all_args);
  free(invcn.relative_paths);
  if(invcn.child_pid) {
    int statloc = -1;
    waitpid(invcn.child_pid, &statloc, 0);
  }
  return 0;
}

File: fortran/fullpath.f90
PROGRAM Fullpath
  IMPLICIT NONE
  INTEGER, PARAMETER :: strlen=2048
  INTEGER :: i, j, print_len, status=0, ipath=1, num_paths=0

  ! Apparently I can't have a string of unknown length, so I'm
  ! just making it large enough to hold most things it could see
  CHARACTER(len=strlen) :: arg, path, dir
  CHARACTER, ALLOCATABLE, DIMENSION(:) :: to_print

  ! Dynamically allocated arrays for holding the paths as we build them up
  CHARACTER(len=strlen), DIMENSION(:), POINTER :: paths, tmp

  LOGICAL :: print_help = .FALSE., copy_output = .FALSE.

  ! Analyze ARGV
  DO i = 1, command_argument_count()
    CALL getarg(i, arg)
    IF (TRIM(arg) == "-h" .OR. TRIM(arg) == "--help") THEN
      print_help = .TRUE.
    ELSE IF (TRIM(arg) == "-c" .OR. TRIM(arg) == "--copy") THEN
      copy_output = .TRUE.
    ELSE IF (LEN_TRIM(arg) /= 0) THEN
      num_paths = num_paths + 1
    END IF
  END DO

  IF (print_help) THEN
    WRITE (*, '(a)') "usage: fullpath *[relative-paths] [-c]"
    WRITE (*, '(a)') ""
    WRITE (*, '(a)') "  Prints the fullpath of the paths"
    WRITE (*, '(a)') "  If no paths are given as args, it will read them from stdin"
    WRITE (*, '(a)') ""
    WRITE (*, '(a)') "  If there is only one path, the trailing newline is omitted"
    WRITE (*, '(a)') ""
    WRITE (*, '(a)') "  The -c flag will copy the results into your pasteboard"
    RETURN
  END IF

  ! Get the paths
  IF (num_paths /= 0) THEN
    ! Copy paths from argv
    allocate(paths(num_paths))
    ipath = 1
    DO i = 1, iargc()
      CALL get_command_argument(i, arg)
      IF (LEN_TRIM(arg) /= 0) THEN
        paths(ipath) = arg
        ipath = ipath + 1
      END IF
    END DO
  ELSE
    ! Read paths from stdin
    num_paths = 0
    allocate(paths(0))
    DO WHILE (.TRUE.)
      ! Reset the path
      DO i=1, strlen
        path(i:i) = achar(32)
      END DO

      ! Read one path
      i=1
      DO WHILE(.TRUE.)
        call fget(path(i:i), status)
        ! break if we hit the end of the stream
        IF (status /= 0) THEN
          EXIT
        END IF
        ! break if we found a newline
        IF (path(i:i) == achar(10)) THEN
          path(i:i) = achar(32) ! remove the newline
          EXIT
        END IF
        i = i + 1
      END DO

      ! If we found a path
      IF (i /= 1) THEN
        ! allocate room for one the existing and new path
        allocate(tmp(num_paths+1))

        ! copy the paths over
        DO ipath=1, num_paths
          tmp(ipath) = paths(ipath)
        END DO

        ! add the path we just read in
        tmp(num_paths+1) = path

        ! deallocate the old paths
        deallocate(paths)

        ! set them to the correct variable
        paths => tmp
        NULLIFY(tmp)

        ! Increase counter
        num_paths = num_paths + 1
      END IF

      ! No more paths, we're done
      IF (status /= 0) THEN
        EXIT
      END IF
    END DO
  END IF


  ! Print the paths
  CALL getcwd(dir)

  IF (num_paths == 1) THEN
    ! Print one path, copy it if we're supposed to
    path = paths(1)
    WRITE(*, '(3a)',advance="no") TRIM(dir), "/", TRIM(path)
    IF (copy_output) THEN
      ! Note that this doesn't work when input has a single-quote in it -.-
      call execute_command_line("printf %s '"//TRIM(dir)//"/"//TRIM(path)//"' | pbcopy")
    END IF
  ELSE
    ! Print all the paths, count how much we printed
    print_len = 0
    DO ipath=1, num_paths
      path = TRIM(dir)//"/"//TRIM(paths(ipath))
      WRITE (*,'(3a)') TRIM(path)
      print_len = LEN_TRIM(path) + 1 ! 1 for the newline
    END DO

    ! Copy the paths if we're supposed to
    ! FIXME: I cannot fucking figure out how to pull this off.
    ! Why doesn't execute_command_line allow me to pass stdin?
    ! (eg allowing me to pass a file descriptor for a pipe would be enough)
    IF (copy_output) THEN
      allocate(to_print(print_len))
      i = 0
      DO ipath=1, num_paths
        path = TRIM(dir)//"/"//TRIM(paths(ipath))//achar(10)
        DO j=1, LEN_TRIM(path)
          to_print(i+j:i+j) = path(j:j)
        END DO
        i = i + j
      END DO
      ! print_len = num_paths*(dirlen+2+strlen) ! dir, slash, path, newline
      ! call execute_command_line("ruby -e 'File.write %(tmp) ARGV[0]'"//to_print//"' | pbcopy")
      ! call execute_command_line("printf %s '"//to_print//"' | pbcopy")
      deallocate(to_print)
    END IF

  END IF

  ! Clean up memory
  deallocate(paths)
END PROGRAM
